### ISSUE TITLE
`time_series_sstable_set::create_single_key_sstable_reader` may return an empty reader even if the queried partition exists (in some other sstable)

### ISSUE THREAD

**Issue by kbr-** [scylladb/scylla#8447]
When a particular partition exists in at least one sstable, the cache
expects any single-partition query to this partition to return a `partition_start`
fragment, even if the result is empty.

In `time_series_sstable_set::create_single_key_sstable_reader` it can
happen that all sstables containing data for the given query get
filtered out and only sstables without the relevant partition are left,
resulting in a reader which immediately returns end-of-stream (while it
should return a `partition_start` and if not in forwarding mode, a
`partition_end`).

**Comment by nyh** [comment]
Is there a test reproducing this bug? Can you write a unit test to reproduce it, and catch possible regressions in this area later?

**Comment by kbr-** [comment]
Funny: the reproducer for this is almost exactly the same as for #8435, you just have to reorder the last two selects.

**Comment by kbr-** [comment]
(and you will then see that the second select returns no data, when it should return a row).

**Comment by kbr-** [comment]
Unfortunately, I haven't been able to reproduce with a unit test. But I do have a dtest (will create a PR):
```diff
diff --git a/cql_additional_tests.py b/cql_additional_tests.py
index c566b28b..f38ad9a2 100644
--- a/cql_additional_tests.py
+++ b/cql_additional_tests.py
@@ -6739,6 +6739,45 @@ class CQLAdditionalTests(Tester):
         out, err = nodes[0].run_cqlsh(cmds='USE veraminetest; DESCRIBE TABLES', show_output=True, return_output=True)
         assert len(out.split()) == 112, 'created 100+ tables'
 
+    # Regression test for scylladb/scylla#8447
+    @attr('single_node')
+    def test_twcs_ck_filtering_with_cache(self):
+        debug('creating single-node cluster')
+        cluster = self.prepare({'ring_delay_ms': 1000})
+        node = cluster.nodelist()[0]
+
+        debug('waiting for connection')
+        session = self.patient_cql_connection(node)
+
+        debug('creating table')
+        session.execute("create keyspace ks with replication = {'class': 'SimpleStrategy', 'replication_factor': 1}")
+        session.execute("create table ks.t (pk int, ck int, primary key (pk, ck))"
+                        " with compaction = {'class': 'TimeWindowCompactionStrategy'} and bloom_filter_fp_chance = 1;")
+
+        debug('creating sstables')
+        session.execute("insert into ks.t (pk, ck) values (1, 0)")
+        node.nodetool('flush')
+        session.execute("insert into ks.t (pk, ck) values (0, 1)")
+        node.nodetool('flush')
+
+        debug('stopping node')
+        node.stop()
+        debug('restarting node')
+        node.start(wait_for_binary_proto=True)
+
+        debug('waiting for connection')
+        session = self.patient_cql_connection(node)
+
+        debug('performing selects')
+        # Both sstables pass through the pk filter (thanks to bloom_filter_fp_chance = 1),
+        # but only the second one - which does not contain the queried partition - passes through the ck filter.
+        # If the created reader does not return `partition_start` (as in #8447)
+        # this will cause the cache to rememeber that there is no partition 1.
+        assert_none(session, "select * from ks.t where pk = 1 and ck = 1")
+        # If the cache remembered that there is no partition 1 in the above query,
+        # this would return no row.
+        assert_one(session, "select * from ks.t where pk = 1 and ck = 0", [1, 0])
+
 
 @attr('dtest-full', 'single_node')
 class MultiColumnRestrictionSimpleTests(Tester):

```

I tried the following unit test which I thought to be equivalent:
```cpp
SEASTAR_TEST_CASE(test_clustering_filtering) {
    auto db_config = make_shared<db::config>();
    db_config->enable_sstables_md_format.set(true);

    return do_with_cql_env_thread([] (cql_test_env& e) {
        cquery_nofail(e, "CREATE TABLE cf(pk int, ck int, PRIMARY KEY(pk, ck)) WITH COMPACTION = {{'class': 'TimeWindowCompactionStrategy'}} AND bloom_filter_fp_chance = 1");
        cquery_nofail(e, "INSERT INTO  cf(pk, ck) VALUES (1, 0)");
        e.db().invoke_on_all([] (database& db) { return db.flush_all_memtables(); }).get();
        cquery_nofail(e, "INSERT INTO  cf(pk, ck) VALUES (0, 1)");
        e.db().invoke_on_all([] (database& db) { return db.flush_all_memtables(); }).get();
        e.db().invoke_on_all([] (database& db) { db.row_cache_tracker().clear(); }).get();

        // Both sstables pass through the pk filter (thanks to bloom_filter_fp_chance = 1),
        // but only the second one - which does not contain the queried partition - passes through the ck filter.
        // If the created reader does not return `partition_start` (as in #8447)
        // this will cause the cache to rememeber that there is no partition 1.
        require_rows(e, "SELECT * FROM cf WHERE pk=1 AND ck=1", {});
        // If the cache remembered that there is no partition 1 in the above query,
        // this would return no row.
        require_rows(e, "SELECT * FROM cf WHERE pk=1 AND ck=0", {{I(1), I(0)}});
    }, cql_test_config(db_config));
}
```
but the second check passes (without my fix) while it should fail.
I'm guessing  that `e.db().invoke_on_all([] (database& db) { db.row_cache_tracker().clear(); }).get();` does not actually clear the cache entirely, so the effect is different than when I restart the node in the dtest.

@tgrabiec do you have an idea why I can't reproduce the bug with this unit test, while the dtest works?

**Comment by tgrabiec** [comment]
@kbr- No clue. `cache_tracker::clear()` should clear the cache. Maybe try printing the contents of the cache (it has an operator<<) to see if data is really evicted.

**Comment by avikivity** [comment]
Bug was introduced and fixed in 4.4, no backports needed.
